index/segment/mem/segment.go

```diff
@@ -22,224 +22,271 @@ package mem
-	if err != nil {
-		return err
-	}
+	{
+		s.writer.Lock()
 
-	if docID != nil {
-		ok, err := s.termsDict.ContainsTerm(doc.IDReservedFieldName, docID)
-		if err != nil {
-			return fmt.Errorf("could not lookup document ID '%s'", docID)
-		}
-		if ok {
-			// Index already contains this document.
-			return nil
-		}
-	} else {
-		uuid, err := s.newUUIDFn()
-		if err != nil {
-			return err
+		b := index.NewBatch([]doc.Document{d})
+		if err := s.prepareDocsWithLocks(b); err != nil {
+			return nil, err
 		}
-		d.Fields = append(d.Fields, doc.Field{
-			Name:  doc.IDReservedFieldName,
-			Value: uuid,
-		})
-	}
 
-	// TODO: Consider supporting concurrent writes by relaxing the requirement that
-	// inserted documents are immediately searchable.
-	s.ids.Lock()
+		// Update the document in case we generated a UUID for it.
+		d = b.Docs[0]
 
-	newID := s.ids.writer.Inc()
-	s.insertDoc(newID, d)
-	err = s.insertTerms(newID, d)
-	s.ids.reader.Inc()
+		s.insertDocWithLocks(d)
+		s.readerID.Inc()
 
-	s.ids.Unlock()
+		s.writer.Unlock()
+	}
 
-	return err
+	return d.ID, nil
 }
 
-func (s *segment) Reader() (index.Reader, error) {
+func (s *segment) InsertBatch(b index.Batch) error {
 	s.state.RLock()
+	defer s.state.RUnlock()
 	if s.state.closed {
-		s.state.RUnlock()
-		return nil, sgmt.ErrClosed
+		return sgmt.ErrClosed
 	}
 
-	maxID := s.ids.reader.Load()
-	r := newReader(s, maxID)
-	return r, nil
-}
+	var err error
+	{
+		s.writer.Lock()
 
-func (s *segment) Seal() error {
-	s.state.Lock()
-	if s.state.sealed {
-		s.state.Unlock()
-		return errSegmentSealed
+		err = s.prepareDocsWithLocks(b)
+		if err != nil && !index.IsBatchPartialError(err) {
+			return err
+		}
+
+		for _, d := range b.Docs {
+			s.insertDocWithLocks(d)
+		}
+		s.readerID.Add(uint32(len(b.Docs)))
+
+		s.writer.Unlock()
 	}
 
-	s.state.sealed = true
-	s.state.Unlock()
-	return nil
+	return err
 }
 
-func (s *segment) Close() error {
-	s.state.Lock()
-	if s.state.closed {
-		s.state.Unlock()
-		return sgmt.ErrClosed
+// prepareDocsWithLocks ensures the given documents can be inserted into the index. It
+// must be called with the state and writer locks.
+func (s *segment) prepareDocsWithLocks(b index.Batch) error {
+	s.writer.idSet.Reset()
+
+	batchErr := index.NewBatchPartialError()
+	for i := 0; i < len(b.Docs); i++ {
+		d := b.Docs[i]
+		if err := d.Validate(); err != nil {
+			if !b.AllowPartialUpdates {
+				return err
+			}
+			batchErr.Add(err, i)
+			continue
+		}
+
+		if d.HasID() {
+			if s.termsDict.ContainsTerm(doc.IDReservedFieldName, d.ID) {
+				// The segment already contains this document so we can remove it from those
+				// we need to index.
+				b.Docs[i], b.Docs[len(b.Docs)] = b.Docs[len(b.Docs)], b.Docs[i]
+				b.Docs = b.Docs[:len(b.Docs)-1]
+
+				// Decrement the loop variable since we just removed this document.
+				i--
+				continue
+			}
+
+			if _, ok := s.writer.idSet.Get(d.ID); ok {
+				if !b.AllowPartialUpdates {
+					return errDuplicateID
```

Comment:
Could you make `errDuplicateID` an exported type. I want to be able to distinguish it from other failures downstream